app/Party.php

```diff
@@ -381,38 +381,173 @@ public function findLatest($limit = 10)
+        // A past event is an event where the end time is less than now.
+        //
+        // We want to use local variables when formatted times to avoid windows where the date/time changes under
+        // our feet between one line of code and the next.
+        $date = date('Y-m-d');
+        $time = date('H:i');
+        $query = $query->undeleted();
+        $query = $query->where('event_date', '<', $date)
+            ->orWhere(function($q2) use ($date, $time)  {
+                $q2->where([
+                    [ 'event_date', '=', "'$date'" ],
+                    [ 'end', '<', "'$time'" ]
+                ]);
+            });
+        return $query;
+    }
+
+    public function scopeFuture($query) {
+        // A future event is an event where the start time is greater than to now.
+        $date = date('Y-m-d');
+        $time = date('H:i');
+        $query = $query->undeleted();
+        $query = $query->whereDate('event_date', '>', "'$date'")
+            ->orWhere(function($q2) use ($date, $time) {
+                $q2->where([
+                               [ 'event_date', '=', "'$date'" ],
+                               [ 'start', '>', "'$time'" ]
+                           ]);
+            });
+        return $query;
+    }
+
+    public function scopeActive($query) {
+        // An active event is an event which has started and not yet finished.
+        $date = date('Y-m-d');
+        $time = date('H:i');
+        $query = $query->undeleted();
+        $query = $query->whereDate('event_date', '=', "'$date'")
+            ->where('start', '<=', "'$time'")
+            ->where('end', '>=', "'$time'");
+        return $query;
+    }
+
+    public function scopeApproved($query) {
+        // wordpress_post_id indicates event approval.
+        $query = $query->undeleted();
+        $query = $query->whereNotNull('wordpress_post_id');
+        return $query;
+    }
+
+    public function scopeHostFor($query, $userids = null) {
+        // Events where this user is a host.
+        $this->defaultUserIds($userids);
+        $query = $query->undeleted();
+        $query = $query->join('events_users AS hf', function ($join) use ($userids) {
+            $join->on('hf.event', '=', 'events.idevents');
+            $join->whereIn('hf.user', $userids);
+            $join->where('hf.role', '=', Role::HOST);
+        })->select('events.*');
+
+        return $query;
+    }
+
+    public function scopeAttendingOrAttended($query, $userids = null) {
+        // Events this user has attending/is attending.
+        $this->defaultUserIds($userids);
+        $query = $query->undeleted();
+        $query = $query->join('events_users AS aoa', function ($join) use ($userids) {
+            $join->on('aoa.event', '=', 'events.idevents');
+            $join->whereIn('aoa.user', $userids);
+
+            // Check the status so that we exclude any events we have been invited to but not confirmed.
+            $join->where('aoa.status', 1);
```

Comment:
I have a vague memory of a bug once, that because `status` is a string, this specific where statement will evaluate to true when the status field is an invite string that starts with 1.  Would be worth a test to check that.

From memory we had to use something like `where('aoa.status', 'like', '1')` to resolve.

Reply:
Thank you.  There is a comment to that effect in PartyController around line 571, so I will duplicate that here.